src/github.com/getlantern/netx/copy.go

```diff
@@ -0,0 +1,92 @@
+package netx
+
+import (
+	"io"
+	"net"
+	"sync/atomic"
+	"time"
+)
+
+var (
+	copyTimeout = 1 * time.Second
+)
+
+// BidiCopy copies between in and out in both directions using the specified
+// buffers, returning the errors from copying to out and copying to in. BidiCopy
+// continues trying to write out to the respective connections for up to
+// writeTimeout to flush any buffered data before giving up and returning
+// io.ErrShortWrite.
+func BidiCopy(out net.Conn, in net.Conn, bufOut []byte, bufIn []byte, writeTimeout time.Duration) (outErr error, inErr error) {
+	stop := uint32(0)
+	outErrCh := make(chan error, 1)
+	inErrCh := make(chan error, 1)
+	go doCopy(out, in, bufIn, writeTimeout, outErrCh, &stop)
+	go doCopy(in, out, bufOut, writeTimeout, inErrCh, &stop)
+	return <-outErrCh, <-inErrCh
+}
+
+// doCopy is based on io.copyBuffer
+func doCopy(dst net.Conn, src net.Conn, buf []byte, writeTimeout time.Duration, errCh chan error, stop *uint32) {
+	var err error
+	defer func() {
+		atomic.StoreUint32(stop, 1)
```

Comment:
@fffw I tested and you were right, setting the deadline only once we're ready to close the connection performs better, and it seems to work on windows, os x and linux.  So, I did it!